### ISSUE TITLE
Emulating a number of instructions does not work as intended

### ISSUE THREAD

**Issue by elicn** [unicorn-engine/unicorn#1606]
When calling `emu_start` with a `count` value, Unicorn seems to emulate more instructions than requested.
For example, when specifying `count=1` Unicorn is expected to perform a single step, but in fact it emulates several instructions.
Using a code hook suggests that the `HOOK_CODE` event might not be dispatched for every instruction, and hence the erroroneous instruction count.

The following Qiling script halts on main function and performs and single step. Output shows that there were several instructions emulated instead of just one. If modified to include a code hook, the hook will be called every several instructions (on basic block boundary...?) instead of every single one.

```python
#!/usr/bin/python3

# run me:
#	PYTHONPATH=/path/to/qiling QL_HOME=/path/to/qiling python3 step.py

import os

from qiling import Qiling

def show_disassembly(ql: Qiling, address: int, count: int) -> None:
	data = ql.mem.read(address, 16 * count)

	for insn in ql.arch.disassembler.disasm(data, address, count):
		marker = '->' if insn.address == address else '  '

		ql.log.warning(f' {marker} {insn.address:#010x} : {insn.bytes.hex():14s} {insn.mnemonic:10s} {insn.op_str}')


if __name__ == '__main__':
	QL_HOME = os.environ.get('QL_HOME', '')

	argv = [fr'{QL_HOME}/examples/rootfs/x86_linux/bin/x86_hello']
	rootfs = fr'{QL_HOME}/examples/rootfs/x86_linux'

	ql = Qiling(argv, rootfs)

	def __hook_main(ql: Qiling):
		# remove the hook so we don't trigger it again
		ql.hook_del(hret)

		# stop emulation at main
		ql.stop()

	hret = ql.hook_address(__hook_main, 0x5655551d)

	ql.log.warning(f'running')
	ql.run()

	# should reach here after hitting main
	pc = ql.arch.regs.arch_pc

	# disassemble several next instructions
	ql.log.warning(f'stopped at {pc:#010x}:')
	show_disassembly(ql, pc, 24)

	steps = 1

	# step
	ql.log.warning(f'performing {steps} steps')
	ql.emu_start(pc, ql.os.exit_point or ql.exit_point, count=steps)

	# should reach here when done stepping
	pc = ql.arch.regs.arch_pc

	# disassemble current instruction
	ql.log.warning(f'stopped at {pc:#010x}:')
	show_disassembly(ql, pc, 1)
```
As shown on the output, emulation is expected to go from 0x5655551d (first instruction in `main`) to 0x56555521 (second instruction in `main`), but stops at 0x5655555c instead.

**Comment by wtdcode** [comment]
The main function is cached when you would like to perform singestep, so you have to clear cache.

```python
#!/usr/bin/python3

# run me:
#    PYTHONPATH=/path/to/qiling QL_HOME=/path/to/qiling python3 step.py

import os

from qiling import Qiling

def show_disassembly(ql: Qiling, address: int, count: int) -> None:
    data = ql.mem.read(address, 16 * count)

    for insn in ql.arch.disassembler.disasm(data, address, count):
        marker = '->' if insn.address == address else '  '

        ql.log.warning(f' {marker} {insn.address:#010x} : {insn.bytes.hex():14s} {insn.mnemonic:10s} {insn.op_str}')


if __name__ == '__main__':
    QL_HOME = os.environ.get('QL_HOME', '')

    argv = [fr'{QL_HOME}/examples/rootfs/x86_linux/bin/x86_hello']
    rootfs = fr'{QL_HOME}/examples/rootfs/x86_linux'

    ql = Qiling(argv, rootfs)

    def __hook_main(ql: Qiling):
        # remove the hook so we don't trigger it again
        ql.hook_del(hret)

        # stop emulation at main
        ql.stop()

    hret = ql.hook_address(__hook_main, 0x5655551d)

    ql.log.warning(f'running')
    ql.run()

    # should reach here after hitting main
    pc = ql.arch.regs.arch_pc

    # disassemble several next instructions
    ql.log.warning(f'stopped at {pc:#010x}:')
    show_disassembly(ql, pc, 24)

    steps = 1

    # step
    ql.log.warning(f'performing {steps} steps')
    ql.uc.ctl_remove_cache(pc, pc + 4096) # <---- Note this line!
    ql.emu_start(pc, ql.os.exit_point or ql.exit_point, count=steps)

    # should reach here when done stepping
    pc = ql.arch.regs.arch_pc

    # disassemble current instruction
    ql.log.warning(f'stopped at {pc:#010x}:')
    show_disassembly(ql, pc, 1)
```

This is not intuitive I admit and might be improved in several ways.

**Comment by elicn** [comment]
Can you elaborate a little bit about that extra line?
What the arguments stand for, and what side effects it may bring?
BTW - it looks like the `ctl_remove_cache` method accepts just one argument.

**Comment by wtdcode** [comment]
Yes, it's a bug fixed recently on dev branch.

**Comment by elicn** [comment]
Confirmed to be fixed on Uc 2.0
